staging/src/k8s.io/apiserver/pkg/util/version/registry.go

```diff
@@ -0,0 +1,433 @@
+		"The versions different components emulate their capabilities (APIs, features, ...) of.\n"+
+		"If set, the component will emulate the behavior of this version instead of the underlying binary version.\n"+
+		"Version format could only be major.minor, for example: '--emulated-version=wardle=1.2,kube=1.31'. Options are:\n"+strings.Join(r.unsafeVersionFlagOptions(true), "\n"))
+
+	fs.Var(cliflag.NewColonSeparatedMultimapStringStringAllowDefaultEmptyKey(&r.featureGatesConfig), "feature-gates", "Comma-separated list of component:key=value pairs that describe feature gates for alpha/experimental features of different components.\n"+
+		"If the component is not specified, defaults to \"kube\". This flag can be repeatedly invoked. For example: --feature-gates 'wardle:featureA=true,wardle:featureB=false' --feature-gates 'kube:featureC=true'"+
+		"Options are:\n"+strings.Join(r.unsafeKnownFeatures(), "\n"))
+}
+
+type componentVersion struct {
+	component string
+	ver       *version.Version
+}
+
+func (r *componentGlobalsRegistry) unsafeSetVersionConfig(config cliflag.ConfigurationMap, isEmulation bool, isFromFlag bool) error {
+	alreadySetComponents := map[string]struct{}{}
+	setQueue := []componentVersion{}
+	versionType := "MinCompatibilityVersion"
+	if isEmulation {
+		versionType = "EmulationVersion"
+	}
+	for comp, verStr := range config {
+		if _, ok := r.componentGlobals[comp]; !ok {
+			return fmt.Errorf("component not registered: %s", comp)
+		}
+		// only components without any dependencies can be set from the flag.
+		if isFromFlag {
+			if isEmulation && r.componentsWithDependentEmulationVersion[comp] {
+				return fmt.Errorf("%s of %s is set by mapping, cannot set it by flag", versionType, comp)
+			}
+			if !isEmulation && r.componentsWithDependentMinCompatibilityVersion[comp] {
+				return fmt.Errorf("%s of %s is set by mapping, cannot set it by flag", versionType, comp)
+			}
+		}
+		ver, err := version.Parse(verStr)
+		if err != nil {
+			return err
+		}
+		if ver.Patch() != 0 {
+			return fmt.Errorf("%s patch version not allowed, got: %s", versionType, verStr)
+		}
+		klog.V(klogLevel).Infof("setting %s %s=%s\n", versionType, comp, ver.String())
+		setQueue = append(setQueue, componentVersion{comp, ver})
+	}
+	for len(setQueue) > 0 {
+		cv := setQueue[0]
+		if _, visited := alreadySetComponents[cv.component]; visited {
+			return fmt.Errorf("setting %s of %s more than once, probably version mapping loop", versionType, cv.component)
+		}
+		alreadySetComponents[cv.component] = struct{}{}
+		setQueue = setQueue[1:]
+		var versionMapping map[string]VersionMapping
+		if isEmulation {
+			r.componentGlobals[cv.component].effectiveVersion.SetEmulationVersion(cv.ver)
+			versionMapping = r.emulationVersionMapping[cv.component]
+		} else {
+			r.componentGlobals[cv.component].effectiveVersion.SetMinCompatibilityVersion(cv.ver)
+			versionMapping = r.minCompatibilityVersionMapping[cv.component]
+		}
```

Comment:
I'm confused. When and why is `isEmulation` false and why is the behavior so different?

Reply:
This is an internal function. `isEmulation = false` when we want to use it for `minCompatibilityVersion`, which will be added in the next release.

Reply:
I suggest not adding this now and saving some pain for when you need it.  This does not look correct to me. Having a `SetVersionConfig` function that sometimes sets emulation version and sometimes sets mincompatibility version is hard to read.

Perhaps you want a `getComponent` function that you can then set different values from.  But the bifurcation here is a problem.

Reply:
removed the bifurcation